replay/models/cql.py

```diff
@@ -0,0 +1,217 @@
+        self.model = CQL_d3rlpy.CQL(**params)
+
+    def _predict(
+        self,
+        log: pyspark.sql.DataFrame,
+        k: int,
+        users: pyspark.sql.DataFrame,
+        items: pyspark.sql.DataFrame,
+        user_features: Optional[pyspark.sql.DataFrame] = None,
+        item_features: Optional[pyspark.sql.DataFrame] = None,
+        filter_seen_items: bool = True,
+    ) -> pyspark.sql.DataFrame:
+
+        users = users.toPandas().to_numpy().flatten()
+        items = items.toPandas().to_numpy().flatten()
+
+        pred = pd.DataFrame()
+        
+        # progress = tqdm.auto.tqdm(users, desc='User')
+        for user in users:
+            matrix = pd.DataFrame({
+                'user_idx': np.repeat(user, len(items)),
+                'item_idx': items
+            })
+            matrix['relevance'] = self.model.predict(matrix.to_numpy())
+            top = matrix.sort_values('relevance', ascending=False).head(k)
+            pred = pd.concat((pred, top))
+
+        return DataPreparator.read_as_spark_df(pred)
+
+    def _fit(
+        self,
+        log: DataFrame,
+        user_features: Optional[DataFrame] = None,
+        item_features: Optional[DataFrame] = None,
+    ) -> None:
+        
+        train = self._prepare_data(log.toPandas(), self.k)
+        self.model.fit(train, n_epochs=self.n_epochs)
+
+    def _prepare_data(self, df, k : int):
+        gb = df.sort_values('timestamp').groupby('user_idx')    
+        list_dfs = [gb.get_group(x) for x in gb.groups]
+        df_s = pd.concat(list_dfs)
+
+        #   reward top-K and later watched movies with 1, other - 0
+        idxs = df_s.sort_values(['relevance', 'timestamp'], ascending=False).groupby('user_idx').head(k).index
+        rewards = np.zeros(len(df_s))
+        rewards[idxs] = 1
+        df_s['rewards'] = rewards
+
+        #   every user has his own episode (for the latest movie terminals has 1, other - 0)
+        user_change = (df_s.user_idx != df_s.user_idx.shift())
+        terminals = np.zeros(len(df_s))
+        terminals[user_change] = 1
+        terminals[0] = 0
+        df_s['terminals'] = terminals
+
+        train_dataset = MDPDataset(
+            observations = np.array(df_s[['user_idx', 'item_idx']]),
+            actions = np.array(df_s['relevance'] +
+                               0.1 * np.random.randn(len(df_s)))[:, None],
+            rewards = df_s['rewards'],
+            terminals = df_s['terminals']
+        )
+        return train_dataset
+    
+    @property
+    def _init_args(self):
+        pass
+
+    def _batch_pass(self, batch, model) -> Dict[str, Any]:
+        pass
+
+    def _loss(self, **kwargs) -> torch.Tensor:
+        pass
+
+    @staticmethod
+    def _predict_by_user(
+        pandas_df: pd.DataFrame,
+        model: nn.Module,
+        items_np: np.ndarray,
+        k: int,
+        item_count: int,
+    ) -> pd.DataFrame:
+        pass
+
+    @staticmethod
+    def _predict_by_user_pairs(
```

Comment:
we need `predict_pairs` method, so when inheriting from `Recommender` please add `_predict_pairs`